### PR TITLE
Starting VirtIO implementation

### PR DESCRIPTION
**Pull request by Quentincestino**
This is the start of a VirtIO library.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -60,3 +60,4 @@ meta/toolchain/gnu/local
 meta/toolchain/gnu/tarballs
 
 docs/brutal.pdf
+sources/libs/virtio/virtio-1.1-spec.pdf
```
Comment (Supercip971) [inline .gitignore:63]: why this ?

File: sources/libs/virtio/block_device.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    struct PACKED virtio_blk_geometry
+    {
+        le_uint16_t cylinders;
+        uint8_t heads;
+        uint8_t sectors;
+    } geometry;
+
+    le_uint32_t block_size;
+
+    struct PACKED virtio_blk_topology
```
Comment (Supercip971) [inline sources/libs/virtio/block_device.h:38]: maybe you should separates these structs or don't put these in struct like:

```c


    le_uint32_t seg_max;

    struct PACKED virtio_blk_geometry
    {
        le_uint16_t cylinders;
        uint8_t heads;
        uint8_t sectors;
    } geometry;

    le_uint32_t block_size;

```

to

```c

    le_uint32_t seg_max;
    le_uint16_t cylinders;
    uint8_t heads;
    uint8_t sectors;

    le_uint32_t block_size;

```
or:
```c


    le_uint32_t seg_max;
   
    VirtioBlkGeometry geometry;

    le_uint32_t block_size;

```

File: sources/libs/virtio/block_device.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    le_uint32_t max_discard_seg;
+    le_uint32_t discard_sector_alignment;
+    le_uint32_t max_write_zeroes_sectors;
+    le_uint32_t max_write_zeroes_seg;
+    uint8_t write_zeroes_way_unmap;
+    uint8_t unused1[3];
+} VirtioBlockConfig;
+
+// VirtIO block device request types
+#define REQUEST_IN 0            // Read
```
Comment (Supercip971) [inline sources/libs/virtio/block_device.h:58]: maybe do an enum ?

File: sources/libs/virtio/device.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <brutal/base.h>
+
+#define VIRTIO_STATUS_ACKNOWLEDGE 0b00000001      // Indicates that the guest OS has found the device and recognized it as a valid virtio device.
+#define VIRTIO_STATUS_DRIVER_AVAILABLE 0b00000010 // Indicates that the guest OS knows how to drive the device.
+#define VIRTIO_STATUS_FAILURE 0b10000000          // Indicates that something went wrong in the guest, and it has given up on the device.
+#define VIRTIO_STATUS_FEATURES_OK 0b00001000      // Indicates that the driver has acknowledged all the features it understands, and feature negotiation is complete.
+#define VIRTIO_STATUS_DRIVER_OK 0b00000100        // Indicates that the driver is set up and ready to drive the device.
+#define VIRTIO_STATUS_NEEDS_RESET 0b01000000      // Indicates that the device has experienced an error from which it can’t recover.
```
Comment (Supercip971) [inline sources/libs/virtio/device.h:10]: maybe do:
```c
#define VIRTIO_STATUS_ACKNOWLEDGE (1 << 0)      // Indicates that the guest OS has found the device and recognized it as a valid virtio device.
#define VIRTIO_STATUS_DRIVER_AVAILABLE (1 << 1)  // Indicates that the guest OS knows how to drive the device.
```
...

File: sources/libs/virtio/entropy_device.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#include <brutal/base.h>
```
Comment (Supercip971) [inline sources/libs/virtio/entropy_device.h:1]: no pragma once ?

**Review by Supercip971 — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (Quentincestino) [reply .gitignore:63]: I have my own copy of the VirtIO spec in order to mark some useful things on it via Foxit Reader, but I guess I shouldn't push it. Maybe you want from me to remove it ?

File: sources/libs/.build.mk [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+DEPENDENCIES += $(LIBS_HOSTED_OBJ:.o=.d)
+
+$(BUILDDIR_HOST)/libs/%.c.o: sources/libs/%.c
+	$(MKCWD)
+	$(HOST_CC) -c -o $@ $< $(HOST_CFLAGS)
+
+$(LIBS_HOSTED_BIN): $(LIBS_HOSTED_OBJ)
+	$(MKCWD)
+	$(HOST_AR) $(HOST_ARFLAGS) $@ $^
```
Comment (Supercip971) [inline sources/libs/.build.mk:53]: this should be removed :/

**Review by Supercip971 — CHANGES_REQUESTED** [review]

**Pull request merged**